unit_tests/src/test_rc_switch_user.cpp

```diff
@@ -0,0 +1,204 @@
+    CHECK(std::strcmp(conn_ptr->clientUser.userName, env.rodsUserName) == 0);
+    CHECK(std::strcmp(conn_ptr->clientUser.rodsZone, env.rodsZone) == 0);
+    // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+}
+
+TEST_CASE("rc_switch_user honors permission model following successful invocation")
+{
+    //
+    // IMPORTANT: This test requires access to a rodsadmin user!
+    //
+
+    load_client_api_plugins();
+
+    rodsEnv env;
+    _getRodsEnv(env);
+
+    const auto sandbox = fs::path{static_cast<const char*>(env.rodsHome)} / "unit_testing_sandbox";
+    irods::experimental::client_connection conn;
+
+    if (!fs::client::exists(conn, sandbox)) {
+        REQUIRE(fs::client::create_collection(conn, sandbox));
+    }
+
+    irods::at_scope_exit remove_sandbox{
+        [&conn, &sandbox] { REQUIRE(fs::client::remove_all(conn, sandbox, fs::remove_options::no_trash)); }};
+
+    // As the administrator, create a data object.
+    const auto data_object = sandbox / "foo";
+    io::client::native_transport tp{conn};
+    io::odstream{tp, data_object} << "data";
+    REQUIRE(fs::client::is_data_object(conn, data_object));
+
+    // Create a test user.
+    const adm::user alice{"test_user_alice"};
+    REQUIRE(adm::client::add_user(conn, alice).value() == 0);
+
+    irods::at_scope_exit remove_test_user{
+        [&conn, &alice] { REQUIRE(adm::client::remove_user(conn, alice).value() == 0); }};
+
+    // Become the test user.
+    auto* conn_ptr = static_cast<RcComm*>(conn);
+    REQUIRE(rc_switch_user(conn_ptr, alice.name.c_str(), alice.zone.c_str()) == 0);
+    // NOLINTBEGIN(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    CHECK(conn_ptr->clientUser.userName == alice.name);
+    CHECK(conn_ptr->clientUser.rodsZone == alice.zone);
+    // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+
+    // Show that the test user cannot see the administrator's collection or data object.
+    CHECK_FALSE(fs::client::exists(conn, sandbox));
+    CHECK_FALSE(fs::client::exists(conn, data_object));
+
+    // Become the administrator so that the test can clean up properly.
+    REQUIRE(rc_switch_user(conn_ptr, env.rodsUserName, env.rodsZone) == 0);
+    CHECK(std::strcmp(conn_ptr->clientUser.userName, env.rodsUserName) == 0);
+    CHECK(std::strcmp(conn_ptr->clientUser.rodsZone, env.rodsZone) == 0);
+    // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+}
+
+TEST_CASE("rc_switch_user enforces API requirements")
+{
+    //
+    // IMPORTANT: This test requires access to a rodsadmin user!
+    //
+
+    load_client_api_plugins();
+
+    rodsEnv env;
+    _getRodsEnv(env);
+
+    irods::experimental::client_connection conn;
+
+    auto* conn_ptr = static_cast<RcComm*>(conn);
+
+    SECTION("null pointers")
+    {
+        REQUIRE(rc_switch_user(conn_ptr, nullptr, nullptr) == SYS_INVALID_INPUT_PARAM);
+        // Become the administrator so that the test can clean up properly.
```

Comment:
Is it necessary to switch back to the admin user in these cases? I thought the switch user call would fail and we would continue to be the admin user.

Reply:
That comment is incorrect. The user will not be changed on an error.

Reply:
Will remove the unnecessary calls to rc_switch_user too.

Reply:
done